libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,312 @@
+// - it is trivially destructible and has a size of less than 40 bytes and an alignment of no more than 8 bytes.
+//
+// "size of" in this case doesn't mean "sizeof". Objects which have padding bytes at the end may also be put into
+// the small buffer if the padding byte can be detected by [[no_unique_address]]. For example
+//
+//   class [[clang::trivial_abi]] A {
+//     int* begin;
+//     int* end;
+//     int* cap;
+//     int flags;
+//   public:
+//     A(A&&);
+//     ~A();
+//     void operator()() {}
+//   };
+//
+// is put into the small buffer, even though sizeof(A) == 32.
+
+template <class...>
+class move_only_function;
+
+template <class...>
+struct __is_move_only_function : false_type {};
+
+template <class _ReturnT, class... _ArgTypes>
+struct __is_move_only_function<move_only_function<_ReturnT(_ArgTypes...)>> : true_type {};
+
+_LIBCPP_END_NAMESPACE_STD
+
+// This header is only partially guarded on purpose. This header is an implementation detail of move_only_function.h
+// and generates multiple versions of std::move_only_function
+#endif // _LIBCPP___FUNCTIONAL_MOVE_ONLY_FUNCTION_IMPL_H
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_CV
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_CV
+#endif
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS _LIBCPP_MOVE_ONLY_FUNCTION_CV&
+#else
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS _LIBCPP_MOVE_ONLY_FUNCTION_CV _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#endif
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT false
+#endif
+
+#define _LIBCPP_MOVE_ONLY_FUNCTION_CVREF _LIBCPP_MOVE_ONLY_FUNCTION_CV _LIBCPP_MOVE_ONLY_FUNCTION_REF
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#ifdef _LIBCPP_ABI_MOVE_ONLY_FUNCTION_TRIVIAL_ABI
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI [[_Clang::__trivial_abi__]]
+#else
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI
+#endif
+
+template <class _ReturnT, class... _ArgTypes>
+class _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI move_only_function<_ReturnT(
+    _ArgTypes...) _LIBCPP_MOVE_ONLY_FUNCTION_CVREF noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT)> {
+  template <class _Functor>
+  struct __function_wrappers_impl {
+    __function_wrappers_impl()                                = delete;
+    __function_wrappers_impl(const __function_wrappers_impl&) = delete;
+    __function_wrappers_impl(__function_wrappers_impl&&)      = delete;
+
+    _LIBCPP_HIDE_FROM_ABI static void __destroy(void* __store) noexcept {
+      if constexpr (!__fits_in_buffer<_Functor>) {
+        std::destroy_at(*static_cast<_Functor**>(__store));
+        ::operator delete(*static_cast<void**>(__store));
+      } else {
+        std::destroy_at(static_cast<_Functor*>(__store));
+      }
+    }
+
+    _LIBCPP_HIDE_FROM_ABI static _ReturnT
+    __call(_LIBCPP_MOVE_ONLY_FUNCTION_CV std::byte* __functor, _ArgTypes... __args) {
```

Comment:
You'll have to use `std::invoke_r` here eventually, in order to deal with `move_only_function<void()>`.

Reply:
I imagine you've already figured it out, but see https://github.com/llvm/llvm-project/commit/67d29a12da3a003b91b9e24ca707228c0316a4e7#diff-ccb2cca347563156b183238a1f19d1ade1677fb178fd17a972a59a6f2b077ad0R99
```
    auto lam = [&]() -> int { called_it = true; return 42; };
    std::move_only_function<void()> f = lam;
    f();
```

Reply:
Yes, I've figured it out. Added a TODO.